flowmachine/flowmachine/features/subscriber/label_event_score.py

```diff
@@ -0,0 +1,192 @@
+    ) -> Dict[str, BaseGeometry]:
+        """
+        Takes a dictionary of labels and bounds expressed as lists of geojson shapes
+        and returns a dictionary of labels and bounds expressed as Shapely polygons.
+
+        Parameters
+        ----------
+        geojson_labels : dict
+            String -> geojson mappings
+        Returns
+        -------
+        dict
+            Dict of labels mapped to lists of shapely polygons
+
+        """
+        bounds_dict = {}
+        for label, geom in geojson_labels.items():
+            try:
+                bounds_dict[label] = shape(geom)
+            except (AttributeError, IndexError, ValueError) as e:
+                raise ValueError(f"Geometry for {label} is not valid: {e}")
+        return bounds_dict
+
+    def _make_query(self):
+
+        scores_cols = self.scores.column_names
+        scores = f"({self.scores.get_query()}) AS scores"
+
+        sql = f"""
+        SELECT COALESCE(score_bounds.label, 'unknown') as label, scores.* FROM {scores}
+        LEFT JOIN {LabelEventScore._get_sql_bound(self.labels)}
+        ON st_contains(score_bounds.geom, st_point(scores.score_hour, scores.score_dow))
+        """
+
+        if self.required is not None:
+            scores_cols = ", ".join([f"labelled.{c}" for c in scores_cols])
+            sql = f"""
+
+                WITH labelled AS ({sql}),
+                    filtered AS (SELECT subscriber AS subscriber FROM labelled
+                                GROUP BY subscriber, label HAVING label = '{self.required}')
+
+                SELECT '{self.required}' AS label, {scores_cols}
+                FROM labelled
+                LEFT JOIN filtered
+                ON labelled.subscriber = filtered.subscriber
+                WHERE filtered.subscriber IS NULL
+                UNION ALL
+                SELECT label, {scores_cols}
+                FROM labelled
+                RIGHT JOIN filtered
+                ON labelled.subscriber = filtered.subscriber
+
+            """
+
+        return sql
+
+    @property
+    def column_names(self) -> List[str]:
+        return ["label"] + self.scores.column_names
+
+    @staticmethod
+    def _get_sql_bound(bounds: Dict[str, List[Dict[str, float]]]) -> str:
+        """
+        Translate a dict of label and lists of score boundary dictionaries into an sql case
+        statement.
+
+        Parameters
+        ----------
+        bounds : dict
+            Dictionary of labels, with lists of score boundaries that should have that label
+
+
+        Returns
+        -------
+        str
+            SQL case statement
+
+        """
+        table_rows = []
+        for label, label_bounds in bounds.items():
+            table_rows.append(f"('{label}', '{label_bounds.wkt}'::geometry)")
+        return f"(VALUES {', '.join(table_rows)}) as score_bounds(label, geom)"
+
+    @staticmethod
+    def bounds_dict_has_overlaps(bounds: Dict[str, BaseGeometry]) -> bool:
```

Comment:
See comment above about renaming this, e.g. to `verify_bounds_dict_has_no_overlaps`?